components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -2,19 +2,277 @@
+    fn callback(&self, _cx: JSContext, _v: HandleValue, _realm: InRealm) {
+        todo!();
+    }
+}
+
+// FIXME: this is similar to [`script::dom::readablestream::ReaderType`] but without reader
+#[derive(JSTraceable)]
+enum ReaderType {
+    BYOB,
+    Default,
+}
+
+// TODO: function?
+#[derive(JSTraceable)]
+enum ViewConstructor {
+    TypedArray,
+    DataView,
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+struct PullIntoDescriptor {
+    buffer: Vec<u8>,
+    buffer_byte_length: usize,
+    byte_offset: usize,
+    byte_length: usize,
+    bytes_filled: usize,
+    minimun_fill: usize,
+    element_size: usize,
+    #[ignore_malloc_size_of = "ViewConstructor"]
+    view_constructor: ViewConstructor,
+    #[ignore_malloc_size_of = "ReaderType"]
+    reader_type: Option<ReaderType>,
+}
 
 /// <https://streams.spec.whatwg.org/#readablebytestreamcontroller>
 #[dom_struct]
 pub struct ReadableByteStreamController {
     reflector_: Reflector,
+
+    /// <https://streams.spec.whatwg.org/#readablebytestreamcontroller-queue>
+    queue: RefCell<QueueWithSizes>,
+
+    underlying_source: Dom<UnderlyingSourceContainer>,
+
+    stream: MutNullableDom<ReadableStream>,
+
+    auto_allocate_chunk_size: Option<usize>,
+
+    pending_pull_intos: RefCell<Vec<PullIntoDescriptor>>,
+}
+
+impl ReadableByteStreamController {
+    pub fn new_inherited(
+        global: &GlobalScope,
+        underlying_source_type: UnderlyingSourceType,
+    ) -> ReadableByteStreamController {
+        ReadableByteStreamController {
+            reflector_: Reflector::new(),
+            queue: RefCell::new(Default::default()),
+            underlying_source: Dom::from_ref(&*UnderlyingSourceContainer::new(
+                global,
+                underlying_source_type,
+            )),
+            stream: MutNullableDom::new(None),
+            auto_allocate_chunk_size: None,
+            pending_pull_intos: RefCell::new(Vec::new()),
+        }
+    }
+
+    pub fn new(
+        global: &GlobalScope,
+        underlying_source: UnderlyingSourceType,
+    ) -> DomRoot<ReadableByteStreamController> {
+        reflect_dom_object(
+            Box::new(ReadableByteStreamController::new_inherited(
+                global,
+                underlying_source,
+            )),
+            global,
+        )
+    }
+
+    pub fn set_stream(&self, stream: &ReadableStream) {
+        self.stream.set(Some(stream))
+    }
+
+    /// <https://streams.spec.whatwg.org/#dequeue-value>
+    fn dequeue_value(&self) -> ReadableByteStreamQueueEntry {
+        let mut queue = self.queue.borrow_mut();
+        queue.dequeue_value()
+    }
+
+    /// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-should-call-pull>
+    fn should_pull(&self) -> bool {
+        // TODO: implement the algo.
+        true
+    }
+
+    /// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-call-pull-if-needed>
+    fn call_pull_if_needed(&self) {
+        if !self.should_pull() {
```

Comment:
I found it really helpful when there is step comment for each block of code e.g

```suggestion
        // step 1 
        if !self.should_pull() {
```

Reply:
That's true :) I'll try to add it up